Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -1811,6 +1778,80 @@ void CGContextEOFillPath(CGContextRef context) {
 }
 #pragma endregion
 
+#pragma region Drawing Operations - CGImage
+
+/**
+ @Status Interoperable
+*/
+void CGContextDrawImage(CGContextRef context, CGRect rect, CGImageRef image) {
+    NOISY_RETURN_IF_NULL(context);
+    RETURN_IF(!image);
+
+    // Obtain the pixel format for the image
+    WICPixelFormatGUID imagePixelFormat = _CGImageGetPixelFormat(image);
+
+    CFRetain(image);
+    woc::unique_cf<CGImageRef> refImage;
+    refImage.reset(image);
+
+    if (!_CGIsValidRenderTargetPixelFormat(imagePixelFormat)) {
+        // convert it to a valid pixelformat
+        refImage.reset(_CGImageCreateCopyWithPixelFormat(image, GUID_WICPixelFormat32bppPBGRA));
+    }
+
+    ComPtr<IWICBitmap> bmap = _CGImageGetImageSource(refImage.get());
+    ComPtr<ID2D1Bitmap> d2dBitmap;
+
+    HRESULT result = context->RenderTarget()->CreateBitmapFromWicBitmap(bmap.Get(), nullptr, &d2dBitmap);
+
+    CGAffineTransform flipImage = CGAffineTransformMakeTranslation(rect.origin.x, rect.origin.y + (rect.size.height / 2.0));
+    flipImage = CGAffineTransformScale(flipImage, 1.0, -1.0);
+    flipImage = CGAffineTransformTranslate(flipImage, -rect.origin.x, -(rect.origin.y + (rect.size.height / 2.0)));
+
+    ComPtr<ID2D1CommandList> commandList;
+    HRESULT hr = __CGContextRenderToCommandList(context,
+                                                _kCGCoordinateModeUserSpace,
+                                                &flipImage,
+                                                &commandList,
+                                                [d2dBitmap, rect](CGContextRef context, ID2D1DeviceContext* deviceContext) {
+                                                    deviceContext->DrawBitmap(d2dBitmap.Get(), __CGRectToD2D_F(rect));
+                                                    return S_OK;
+                                                });
+
+    FAIL_FAST_IF_FAILED(__CGContextRenderImage(context, commandList.Get()));
+}
+
+void CGContextDrawImageRect(CGContextRef context, CGImageRef image, CGRect src, CGRect dst) {
+    NOISY_RETURN_IF_NULL(context);
+    RETURN_IF(!image);
+
+    // Default rect
+    CGRect drawRect = CGRectMake(0, 0, CGImageGetWidth(image), CGImageGetHeight(image));
+
+    if (!CGRectEqualToRect(src, dst)) {
+        float scaleX = CGRectGetWidth(dst) / CGRectGetWidth(src);
+        float scaleY = CGRectGetHeight(dst) / CGRectGetHeight(src);
+        float deltaX = CGRectGetMinX(dst) - (CGRectGetMinX(src) * scaleX);
+        float deltaY = CGRectGetMinY(dst) - (CGRectGetMinY(src) * scaleY);
+        drawRect = CGRectMake(deltaX, deltaY, (float)CGImageGetWidth(image) * scaleX, (float)CGImageGetHeight(image) * scaleY);
+    }
+
+    CGContextSaveGState(context);
+    CGContextClipToRect(context, dst);
+    CGContextDrawImage(context, drawRect, image);
```

Comment:
I think there's a d2d drawing mode that will do this in one go. Layer clipping is somewhat expensive!